diff --git a/build/soong/scripts/check_boot_jars/package_allowed_list.txt b/build/soong/scripts/check_boot_jars/package_allowed_list.txt
index 3dc9847..f94b349 100644
--- a/build/soong/scripts/check_boot_jars/package_allowed_list.txt
+++ b/build/soong/scripts/check_boot_jars/package_allowed_list.txt
@@ -246,6 +246,9 @@ com\.google\.i18n\.phonenumbers
 # Packages used for Android in Chrome OS
 org\.chromium\.arc
 org\.chromium\.arc\..*
+####
+com\.elotouch\.library
+com\.elotouch\.library\..*
 
 # QC adds
 com.qualcomm.qti
diff --git a/device/qcom/qssi/base.mk b/device/qcom/qssi/base.mk
index 51be605..67d2e34 100755
--- a/device/qcom/qssi/base.mk
+++ b/device/qcom/qssi/base.mk
@@ -44,6 +44,13 @@ BOARD_HAVE_QCOM_FM ?= true
 #ifeq ($(strip $(TARGET_USES_QTIC_EXTENSION)),true)
 #PRODUCT_BOOT_JARS += com.qualcomm.qti.camera
 #endif
+#<!-- moidfy TaskId20196
+PRODUCT_PACKAGES += eloTouchManager
+#PRODUCT_BOOT_JARS += eloTouchManager
+PRODUCT_PACKAGES += eloTouchService
+#PRODUCT_BOOT_JARS += eloTouchService
+
+#moidfy TaskId20196-->
 
 #skip boot jars check
 SKIP_BOOT_JARS_CHECK := true
diff --git a/frameworks/base/Android.bp b/frameworks/base/Android.bp
index 2f34f8c..bca18e4 100755
--- a/frameworks/base/Android.bp
+++ b/frameworks/base/Android.bp
@@ -356,6 +356,7 @@ java_defaults {
         // If MimeMap ever becomes its own APEX, then this dependency would need to be removed
         // in favor of an API stubs dependency in java_library "framework" below.
         "mimemap",
+	"eloTouchManager",
         "av-types-aidl-java",
         "tv_tuner_resource_manager_aidl_interface-java",
         "soundtrigger_middleware-aidl-java",
diff --git a/frameworks/base/core/java/android/app/SystemServiceRegistry.java b/frameworks/base/core/java/android/app/SystemServiceRegistry.java
old mode 100644
new mode 100755
index 32ea41b..73682ca
--- a/frameworks/base/core/java/android/app/SystemServiceRegistry.java
+++ b/frameworks/base/core/java/android/app/SystemServiceRegistry.java
@@ -233,6 +233,10 @@ import com.android.internal.net.INetworkWatchlistManager;
 import com.android.internal.os.IDropBoxManagerService;
 import com.android.internal.policy.PhoneLayoutInflater;
 import com.android.internal.util.Preconditions;
+//<!-- modify TaskId20196
+import com.elotouch.library.EloGpioManager;
+import com.elotouch.library.IEloGpioManager;
+//modify TaskId20196-->
 
 import java.util.Map;
 import java.util.Objects;
@@ -266,6 +270,17 @@ public final class SystemServiceRegistry {
     private SystemServiceRegistry() { }
 
     static {
+		//<!-- modify TaskId20196
+        registerService(Context.GPIO_SERVICE,EloGpioManager.class,  
+                new CachedServiceFetcher<EloGpioManager>(){  
+            @Override   
+            public EloGpioManager createService(ContextImpl ctx)  
+            {  
+                IBinder b = ServiceManager.getService(Context.GPIO_SERVICE);  
+                Log.i("add_service_test","SystemServiceRegistry registerService method");  
+                return new EloGpioManager(IEloGpioManager.Stub.asInterface(b));  
+            }});
+			//modify TaskId20196-->
         //CHECKSTYLE:OFF IndentationCheck
         registerService(Context.ACCESSIBILITY_SERVICE, AccessibilityManager.class,
                 new CachedServiceFetcher<AccessibilityManager>() {
diff --git a/frameworks/base/core/java/android/content/Context.java b/frameworks/base/core/java/android/content/Context.java
old mode 100644
new mode 100755
index c3ec094..8671c35
--- a/frameworks/base/core/java/android/content/Context.java
+++ b/frameworks/base/core/java/android/content/Context.java
@@ -3711,6 +3711,7 @@ public abstract class Context {
             //@hide: SPEECH_RECOGNITION_SERVICE,
             UWB_SERVICE,
             MEDIA_METRICS_SERVICE,
+	    GPIO_SERVICE,//<!--modify TaskId20196 -->
     })
     @Retention(RetentionPolicy.SOURCE)
     public @interface ServiceName {}
@@ -4343,6 +4344,11 @@ public abstract class Context {
     /** {@hide} */
     public static final String NETWORK_WATCHLIST_SERVICE = "network_watchlist";
 
+	//<!-- modify TaskId20196
+    /** {@hide} */
+    public static final String GPIO_SERVICE= "gpio_service";
+	//modify TaskId20196 -->
+	
     /**
      * Use with {@link #getSystemService(String)} to retrieve a {@link
      * android.net.wifi.WifiManager} for handling management of
diff --git a/frameworks/base/services/Android.bp b/frameworks/base/services/Android.bp
old mode 100644
new mode 100755
index c83a697..7673eda
--- a/frameworks/base/services/Android.bp
+++ b/frameworks/base/services/Android.bp
@@ -125,6 +125,7 @@ java_library {
         "service-blobstore",
         "service-jobscheduler",
         "android.hidl.base-V1.0-java",
+	"eloTouchService",
     ],
 
     libs: [
diff --git a/frameworks/base/services/java/com/android/server/SystemServer.java b/frameworks/base/services/java/com/android/server/SystemServer.java
old mode 100644
new mode 100755
index db6323c..412c341
--- a/frameworks/base/services/java/com/android/server/SystemServer.java
+++ b/frameworks/base/services/java/com/android/server/SystemServer.java
@@ -225,6 +225,9 @@ import java.util.Timer;
 import java.util.TreeSet;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Future;
+//<!-- modify TaskId20196
+import com.android.elotouchservice.EloGpioService;
+//modify TaskId20196 -->
 
 /**
  * Entry point to {@code system_server}.
@@ -1576,6 +1579,14 @@ public final class SystemServer implements Dumpable {
             mSystemServiceManager.startService(AppIntegrityManagerService.class);
             t.traceEnd();
 
+	    //<!-- modify TaskId20196
+            t.traceBegin("GpioService start");
+            EloGpioService mEloGpioService = new EloGpioService(context);    
+            ServiceManager.addService(Context.GPIO_SERVICE, mEloGpioService); 
+            Slog.i("add_service_test", "SystemServer add service");
+            t.traceEnd();
+	    //modify TaskId20196-->
+
         } catch (Throwable e) {
             Slog.e("System", "******************************************");
             Slog.e("System", "************ Failure starting core service");
diff --git a/system/sepolicy/prebuilts/api/26.0/private/service_contexts b/system/sepolicy/prebuilts/api/26.0/private/service_contexts
old mode 100644
new mode 100755
index ff97d66..7e201d9
--- a/system/sepolicy/prebuilts/api/26.0/private/service_contexts
+++ b/system/sepolicy/prebuilts/api/26.0/private/service_contexts
@@ -5,6 +5,9 @@ alarm                                     u:object_r:alarm_service:s0
 android.os.UpdateEngineService            u:object_r:update_engine_service:s0
 android.security.keystore                 u:object_r:keystore_service:s0
 android.service.gatekeeper.IGateKeeperService    u:object_r:gatekeeper_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/prebuilts/api/26.0/public/service.te b/system/sepolicy/prebuilts/api/26.0/public/service.te
old mode 100644
new mode 100755
index da540db..4ab9950
--- a/system/sepolicy/prebuilts/api/26.0/public/service.te
+++ b/system/sepolicy/prebuilts/api/26.0/public/service.te
@@ -145,3 +145,6 @@ type wifi_service, app_api_service, system_server_service, service_manager_type;
 type wificond_service, service_manager_type;
 type wifiaware_service, app_api_service, system_server_service, service_manager_type;
 type window_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
diff --git a/system/sepolicy/prebuilts/api/27.0/private/service_contexts b/system/sepolicy/prebuilts/api/27.0/private/service_contexts
old mode 100644
new mode 100755
index a82243f..191144f
--- a/system/sepolicy/prebuilts/api/27.0/private/service_contexts
+++ b/system/sepolicy/prebuilts/api/27.0/private/service_contexts
@@ -5,6 +5,9 @@ alarm                                     u:object_r:alarm_service:s0
 android.os.UpdateEngineService            u:object_r:update_engine_service:s0
 android.security.keystore                 u:object_r:keystore_service:s0
 android.service.gatekeeper.IGateKeeperService    u:object_r:gatekeeper_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/prebuilts/api/27.0/public/service.te b/system/sepolicy/prebuilts/api/27.0/public/service.te
old mode 100644
new mode 100755
index e97b864..e1418e2
--- a/system/sepolicy/prebuilts/api/27.0/public/service.te
+++ b/system/sepolicy/prebuilts/api/27.0/public/service.te
@@ -148,3 +148,6 @@ type wifi_service, app_api_service, system_server_service, service_manager_type;
 type wificond_service, service_manager_type;
 type wifiaware_service, app_api_service, system_server_service, service_manager_type;
 type window_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
diff --git a/system/sepolicy/prebuilts/api/28.0/private/service_contexts b/system/sepolicy/prebuilts/api/28.0/private/service_contexts
old mode 100644
new mode 100755
index 5ec45a2..7d3193f
--- a/system/sepolicy/prebuilts/api/28.0/private/service_contexts
+++ b/system/sepolicy/prebuilts/api/28.0/private/service_contexts
@@ -5,6 +5,9 @@ alarm                                     u:object_r:alarm_service:s0
 android.os.UpdateEngineService            u:object_r:update_engine_service:s0
 android.security.keystore                 u:object_r:keystore_service:s0
 android.service.gatekeeper.IGateKeeperService    u:object_r:gatekeeper_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/prebuilts/api/28.0/public/service.te b/system/sepolicy/prebuilts/api/28.0/public/service.te
old mode 100644
new mode 100755
index 3526049..b03b29c
--- a/system/sepolicy/prebuilts/api/28.0/public/service.te
+++ b/system/sepolicy/prebuilts/api/28.0/public/service.te
@@ -159,3 +159,6 @@ type wificond_service, service_manager_type;
 type wifiaware_service, app_api_service, system_server_service, service_manager_type;
 type window_service, system_api_service, system_server_service, service_manager_type;
 type wpantund_service, system_api_service, service_manager_type;
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
diff --git a/system/sepolicy/prebuilts/api/29.0/private/service_contexts b/system/sepolicy/prebuilts/api/29.0/private/service_contexts
old mode 100644
new mode 100755
index 96d553b..b45e6c7
--- a/system/sepolicy/prebuilts/api/29.0/private/service_contexts
+++ b/system/sepolicy/prebuilts/api/29.0/private/service_contexts
@@ -12,6 +12,9 @@ app_prediction                            u:object_r:app_prediction_service:s0
 apexservice                               u:object_r:apex_service:s0
 ashmem_device_service                     u:object_r:ashmem_device_service:s0
 gsiservice                                u:object_r:gsi_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/prebuilts/api/29.0/public/service.te b/system/sepolicy/prebuilts/api/29.0/public/service.te
old mode 100644
new mode 100755
index a2193d0..c1a15c6
--- a/system/sepolicy/prebuilts/api/29.0/public/service.te
+++ b/system/sepolicy/prebuilts/api/29.0/public/service.te
@@ -192,6 +192,10 @@ type wpantund_service, system_api_service, service_manager_type;
 ### Neverallow rules
 ###
 
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
+
 # servicemanager handles registering or looking up named services.
 # It does not make sense to register or lookup something which is not a service.
 # Trigger a compile error if this occurs.
diff --git a/system/sepolicy/prebuilts/api/30.0/private/service_contexts b/system/sepolicy/prebuilts/api/30.0/private/service_contexts
old mode 100644
new mode 100755
index 5c6f1a4..c51d0ce
--- a/system/sepolicy/prebuilts/api/30.0/private/service_contexts
+++ b/system/sepolicy/prebuilts/api/30.0/private/service_contexts
@@ -23,6 +23,9 @@ app_search                                u:object_r:app_search_service:s0
 apexservice                               u:object_r:apex_service:s0
 blob_store                                u:object_r:blob_store_service:s0
 gsiservice                                u:object_r:gsi_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/prebuilts/api/30.0/public/service.te b/system/sepolicy/prebuilts/api/30.0/public/service.te
old mode 100644
new mode 100755
index f27772e..85dd3ea
--- a/system/sepolicy/prebuilts/api/30.0/public/service.te
+++ b/system/sepolicy/prebuilts/api/30.0/public/service.te
@@ -216,6 +216,9 @@ type hal_power_service, vendor_service, service_manager_type;
 type hal_rebootescrow_service, vendor_service, service_manager_type;
 type hal_vibrator_service, vendor_service, service_manager_type;
 
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
 ###
 ### Neverallow rules
 ###
diff --git a/system/sepolicy/prebuilts/api/31.0/private/service_contexts b/system/sepolicy/prebuilts/api/31.0/private/service_contexts
old mode 100644
new mode 100755
index 3fd342b..91d3f68
--- a/system/sepolicy/prebuilts/api/31.0/private/service_contexts
+++ b/system/sepolicy/prebuilts/api/31.0/private/service_contexts
@@ -50,6 +50,9 @@ app_search                                u:object_r:app_search_service:s0
 apexservice                               u:object_r:apex_service:s0
 blob_store                                u:object_r:blob_store_service:s0
 gsiservice                                u:object_r:gsi_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/prebuilts/api/31.0/public/service.te b/system/sepolicy/prebuilts/api/31.0/public/service.te
old mode 100644
new mode 100755
index ba7837d..9256a44
--- a/system/sepolicy/prebuilts/api/31.0/public/service.te
+++ b/system/sepolicy/prebuilts/api/31.0/public/service.te
@@ -243,6 +243,10 @@ type wpantund_service, system_api_service, service_manager_type;
 type tethering_service, app_api_service, ephemeral_app_api_service, system_server_service, service_manager_type;
 type emergency_affordance_service, system_server_service, service_manager_type;
 
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
+
 ###
 ### HAL Services
 ###
diff --git a/system/sepolicy/private/service_contexts b/system/sepolicy/private/service_contexts
old mode 100644
new mode 100755
index 3fd342b..91d3f68
--- a/system/sepolicy/private/service_contexts
+++ b/system/sepolicy/private/service_contexts
@@ -50,6 +50,9 @@ app_search                                u:object_r:app_search_service:s0
 apexservice                               u:object_r:apex_service:s0
 blob_store                                u:object_r:blob_store_service:s0
 gsiservice                                u:object_r:gsi_service:s0
+#<!-- modify Task20196 add system service start-->
+gpio_service				  	  u:object_r:sys_gpio_api_service:s0
+#<!-- modify Task20196 add system service end-->
 appops                                    u:object_r:appops_service:s0
 appwidget                                 u:object_r:appwidget_service:s0
 assetatlas                                u:object_r:assetatlas_service:s0
diff --git a/system/sepolicy/public/service.te b/system/sepolicy/public/service.te
old mode 100644
new mode 100755
index ba7837d..9256a44
--- a/system/sepolicy/public/service.te
+++ b/system/sepolicy/public/service.te
@@ -243,6 +243,10 @@ type wpantund_service, system_api_service, service_manager_type;
 type tethering_service, app_api_service, ephemeral_app_api_service, system_server_service, service_manager_type;
 type emergency_affordance_service, system_server_service, service_manager_type;
 
+#<!-- modify Task20196 add system service start-->
+type sys_gpio_api_service, system_api_service, system_server_service, service_manager_type;
+#<!-- modify Task20196 add system service end-->
+
 ###
 ### HAL Services
 ###
